File: src/main/java/kernitus/plugin/OldCombatMechanics/updater/VersionChecker.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package kernitus.plugin.OldCombatMechanics.updater;

import kernitus.plugin.OldCombatMechanics.OCMMain;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class VersionChecker {

    public static boolean shouldUpdate(String remoteVersion) {
        return isUpdateOut(remoteVersion, OCMMain.getVersion());
    }

    private static boolean isUpdateOut(String remoteVersion, String localVersion) {
        final int[] testVer = getVersionNumbers(remoteVersion);
        final int[] baseVer = getVersionNumbers(localVersion);

        for (int i = 0; i < testVer.length; i++) {
            if (testVer[i] != baseVer[i])
                return testVer[i] > baseVer[i];
        }

        return false;
    }

    private static int[] getVersionNumbers(String ver) {
        // Support both -beta and -SNAPSHOT
        Matcher m = Pattern
                .compile("(\\d+)\\.(\\d+)(?:\\.(\\d+))?(?:-(beta|SNAPSHOT)(\\d*))?", Pattern.CASE_INSENSITIVE)
                .matcher(ver);
        if (!m.matches())
            throw new IllegalArgumentException("Plugin version formatted wrong!");

        return new int[] {
                Integer.parseInt(m.group(1)), // MAJOR
                Integer.parseInt(m.group(2)), // MINOR
                m.group(3) == null ? 0 : Integer.parseInt(m.group(3)), // PATCH (default 0)
                m.group(4) == null ? Integer.MAX_VALUE : // Release version
                        (m.group(5).isEmpty() ? 0 : Integer.parseInt(m.group(5))) // Pre-release number
        };
    }

}

File: src/main/kotlin/kernitus/plugin/OldCombatMechanics/updater/VersionChecker.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package kernitus.plugin.OldCombatMechanics.updater

import kernitus.plugin.OldCombatMechanics.OCMMain
import java.util.regex.Pattern

object VersionChecker {
    fun shouldUpdate(remoteVersion: String): Boolean {
        return isUpdateOut(remoteVersion, OCMMain.version)
    }

    private fun isUpdateOut(remoteVersion: String, localVersion: String): Boolean {
        val testVer = getVersionNumbers(remoteVersion)
        val baseVer = getVersionNumbers(localVersion)

        for (i in testVer.indices) {
            if (testVer[i] != baseVer[i]) return testVer[i] > baseVer[i]
        }

        return false
    }

    private fun getVersionNumbers(ver: String): IntArray {
        val m = Pattern.compile("(\\d+)\\.(\\d+)\\.*(\\d*)(-beta(\\d*))?", Pattern.CASE_INSENSITIVE).matcher(ver)
        require(m.matches()) { "Plugin version formatted wrong!" }

        // Group 1 = MAJOR
        // Group 2 = MINOR
        // Group 3 = PATCH
        // Group 4 = beta
        // Group 5 = beta_version

        //This parses it to MAJOR.MINOR.PATCH-beta_version
        //MAJOR & MINOR required, anything else is set to maximum value if omitted - necessary otherwise
        // somebody with a beta version will not see update to release version
        return intArrayOf(
            m.group(1).toInt(),
            m.group(2).toInt(),
            if (m.group(3).isEmpty()) Int.MAX_VALUE else m.group(3).toInt(),
            if (m.group(4) == null) Int.MAX_VALUE else (if (m.group(5).isEmpty()) 1 else m.group(5).toInt())
        )
    }
}
